### PR TITLE
Add support for configuring Envoys route idle_timeout  

### PR DESCRIPTION
**Pull request by oulman**
### Description

This PR adds support for configuring the Envoy [route level idle_timeout](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routeaction-idle-timeout) for service-routers and local_app. For our teams scenario, this will let us extend the timeout for long-running gRPC services so that Envoy doesn't terminate the connection prematurely. 

For service-routers a `IdleTimeout` parameter has been added, and for local_app a proxy config `local_idle_timeout_ms` parameter was created. I followed the [RequestTimeout](https://www.consul.io/docs/connect/config-entries/service-router#requesttimeout) and [local_request_timeout_ms](https://www.consul.io/docs/connect/proxies/envoy#local_request_timeout_ms) definitions closely.

##### service-router IdleTimeout
Created a service-router and configured the `IdleTimeout` to 60s

```
Kind = "service-router"
Name = "waitserver"
Routes = [
  {
    Destination {
      Service = ""
      RequestTimeout = "60m"
      IdleTimeout = "60s"
    }
  }
]
```

Dumped the Envoy config and verified that the `idle_timeout` configuration parameter was set

```
<snip>
        "name": "waitserver",
        "domains": [
         "*"
        ],
        "routes": [
         {
          "match": {
           "prefix": "/"
          },
          "route": {
           "cluster": "waitserver.default.dc1.internal.ba4be18f-a3c7-7e2f-eff4-a25a8a1fce57.consul",
           "timeout": "3600s",
           "idle_timeout": "60s"
          }
         },
<snip>
```

##### local_idle_timeout_ms

Configure `local_idle_timeout_ms` on a service proxy.config

```
service {
  name = "waitserver"
  id = "waitserver"
  address = "10.5.0.20"
  port = 9222

  connect {
    sidecar_service {
      port = 10201

      proxy {
        config {
          local_request_timeout_ms = 0
          local_idle_timeout_ms = 0
        }
      }
      check {
        name = "Connect Envoy Sidecar"
        tcp = "10.5.0.20:10201"
        interval ="10s"
      }
    }
  }
}
```

Dumped the Envoy config and verified that the `idle_timeout` configuration parameter was set on the public_listener route for `local_app`

```
{
     "route_config": {
      "@type": "type.googleapis.com/envoy.config.route.v3.RouteConfiguration",
      "name": "public_listener",
      "virtual_hosts": [
       {
        "name": "public_listener",
        "domains": [
         "*"
        ],
        "routes": [
         {
          "match": {
           "prefix": "/"
          },
          "route": {
           "cluster": "local_app",
           "timeout": "0s",
           "idle_timeout": "0s"
          }
         }
        ]
       }
      ]
     },
```

### Links
Include any links here that might be helpful for people reviewing your PR (Tickets, GH issues, API docs, external benchmarks, tools docs, etc). If there are none, feel free to delete this section.

Please be mindful not to leak any customer or confidential information. HashiCorp employees may want to use our internal URL shortener to obfuscate links.

### PR Checklist

* [ ] updated test coverage
* [ ] external facing docs updated
* [ ] not a security concern

### CODE REVIEW

File: website/content/docs/connect/proxies/envoy.mdx [hunk 1]
```diff
@@ -253,6 +253,11 @@ defaults that are inherited by all services.
   specified, inherits the Envoy default for route timeouts (15s). A value of 0 will
   disable request timeouts.
 
+- `local_idle_timeout_ms` - In milliseconds, the stream timeout for HTTP requests
+  to the local application instance. Applies to HTTP based protocols only. If not
+  specified, inherits the Envoy default for route idle timeouts (300s). A value of 0 will
+  disable request timeouts.
```
[the remark this answers is not in the document]
Comment (oulman) [reply website/content/docs/connect/proxies/envoy.mdx]: Updated. Thank you!

**Comment by oulman** [comment]
@blake can you remove the stale label so this doesn't get closed? 🙏 Thanks!!

**Comment by jkirschner-hashicorp** [comment]
Done. I made this PR staleproof for now!

**Comment by oulman** [comment]
👋  Hi. I wanted to see now that 1.14 is out if this PR could be considered? Thanks!

File: website/content/docs/connect/proxies/envoy.mdx [hunk 2]
```diff
@@ -339,6 +344,7 @@ defaults that are inherited by all services.
   - `exact_balance` - Inbound connections to the service use the
   [Envoy Exact Balance Strategy.](https://cloudnative.to/envoy/api-v3/config/listener/v3/listener.proto.html#config-listener-v3-listener-connectionbalanceconfig-exactbalance)
 
+
```
Comment (dhiaayachi) [inline website/content/docs/connect/proxies/envoy.mdx:347]:
```suggestion
```

File: website/content/docs/connect/proxies/envoy.mdx [hunk 3]
```diff
@@ -331,6 +331,11 @@ defaults that are inherited by all services.
   specified, inherits the Envoy default for route timeouts (15s). A value of 0 will
   disable request timeouts.
 
+- `local_idle_timeout_ms` - In milliseconds, the stream timeout for HTTP requests
+  to the local application instance. Applies to HTTP based protocols only. If not
+  specified, inherits the Envoy default for route idle timeouts (300s). A value of 0 
+  disables request timeouts.
+
```
Comment (dhiaayachi) [inline website/content/docs/connect/proxies/envoy.mdx:338]:
```suggestion
- `local_idle_timeout_ms` - In milliseconds, the idle timeout for HTTP requests
  to the local application instance. Applies to HTTP based protocols only. If not
  specified, inherits the Envoy default for route idle timeouts (15s). A value of 0 
  disables request timeouts.

```

**Review by dhiaayachi — CHANGES_REQUESTED** [review]
Thank you for working on this @oulman. I have a few comments around doc changes and I believe this is missing a test around the `local_request_timeout_ms` at the xds generation level.

**Comment by oulman** [comment]
>  and I believe this is missing a test around the local_request_timeout_ms at the xds generation level.

Thanks, @dhiaayachi! Is that tests in additions to the changes here?

https://github.com/hashicorp/consul/pull/14340/files#diff-fb5ad005250adf2da36018fec648de9d65d05df9de42eed95310bac480e9c0b4

**Comment by dhiaayachi** [comment]
> > and I believe this is missing a test around the local_request_timeout_ms at the xds generation level.
> 
> Thanks, @dhiaayachi! Is that tests in addition to the changes here?
> 
> https://github.com/hashicorp/consul/pull/14340/files#diff-fb5ad005250adf2da36018fec648de9d65d05df9de42eed95310bac480e9c0b4

The missing test would be on the rendering side, could be an addition to [this test](http://github.com/hashicorp/consul/blob/b856a17cbf518ee1f03ea6902e9f344ce5ff3df8/agent/xds/listeners_test.go#L230-L230)

**Comment by oulman** [comment]
> The missing test would be on the rendering side, could be an addition to [this test](https://github.com/hashicorp/consul/blob/b856a17cbf518ee1f03ea6902e9f344ce5ff3df8/agent/xds/listeners_test.go#L230-L230)

Ah! Thank you! I have added the `local_idle_timeout_ms` to that test [in this commit](https://github.com/hashicorp/consul/pull/14340/commits/e078ad21c4ed8057f3887193be385b474f676cfa).

**Comment by oulman** [comment]
> The missing test would be on the rendering side, could be an addition to [this test](https://github.com/hashicorp/consul/blob/b856a17cbf518ee1f03ea6902e9f344ce5ff3df8/agent/xds/listeners_test.go#L230-L230)

Ah! Thank you, @dhiaayach!i I have added the `local_idle_timeout_ms` to that test [in this commit](https://github.com/hashicorp/consul/pull/14340/commits/e078ad21c4ed8057f3887193be385b474f676cfa).

**Comment by david-yu** [comment]
Hi @dhiaayachi I just checked with @oulman and he has confirmed that he has addressed your comments. Are additional changes needed here?

**Review by dhiaayachi — APPROVED** [review]
LGTM! it would be good to have @hashicorp/consul-docs review the documentation bits.

**Pull request merged**